Source/DafnyCore/Resolver/Resolver.cs

```diff
@@ -9742,7 +9742,22 @@ public class ResolveTypeReturn {
           ExprRhs rr = (ExprRhs)s.Rhs;
           ResolveExpression(rr.Expr, resolutionContext);
           Contract.Assert(rr.Expr.Type != null);  // follows from postcondition of ResolveExpression
-          AddAssignableConstraint(stmt.Tok, lhsType, rr.Expr.Type, "RHS (of type {1}) not assignable to LHS (of type {0})");
+
+          bool isHiddenUpdateToOutVar = false;
+          if (s.Lhs is IdentifierExpr) {
+            IdentifierExpr ie = (IdentifierExpr)s.Lhs;
+            if (ie.Var is Formal) {
+              Formal f = ((Formal)ie.Var);
+              if (!f.InParam) {
+                isHiddenUpdateToOutVar = true;
```

Comment:
Based on your PR description, the update is not necessarily hidden, it might be in plain sight.

But I would love to see your new custom message only for implicit updates. We already have a partial mechanism for that in place. Here is how I suggest you could do this:
* Leverage IsImplicit from expressions, which is currently applied only for "this" and static receivers. 
* Make this property writable on IdentifierExpr
* Make Implicit set to true when building the implicit assignment:
https://github.com/dafny-lang/dafny/blob/50c2c99822f7ebe75ce23bab13d207dce5f57633/Source/DafnyCore/Resolver/Resolver.cs#L9559-L9565
* Use this implicit to refine your error message to only the case where the identifier is implicit.

Reply:
Nice!  I went looking for something like `IsImplicit` but missed it, and this is certainly nicer than the best idea I could come up, which was to extend a subclass for `IdentifierExpr`.  Will do so now, thanks for the pointers!

Reply:
Done - as of 59991658c64ea9348b2fed19e2505a773e02ef16 the resolver distingushes implicit assignments (the unit test is updated to make this distinction explicit)